File: sparse-matrix-multiplication.cpp
//Time - O(row(A)*cols(A)*cols(B))
//Space - O(row(A)*cols(B))
class Solution {
  public:
    vector<vector<int>> sparseMatrixMultiplication(vector<vector<int>> A, vector<vector<int>> B){
        int Ar = A.size();
        int Ac = A[0].size();
        int Br = B.size();
        int Bc = B[0].size();
        
        vector<vector<int>> result (Ar,vector<int>(Bc,0));
        
        for(int i = 0;i<Ar;i++){
            for(int k = 0;k<Ac;k++){
                if(A[i][k] != 0){
                    for(int j = 0;j<Bc;j++){
                        if(B[k][j]!=0){
                            result[i][j] += A[i][k]*B[k][j];
                        }
                    }
                }
            }
        }
        return result;
    }
};

int main() {
    std::cout << "Hello World!\n";
    Solution sol;
    vector<vector<int>> A {{1,0,0},{-1,0,3}};
    vector<vector<int>> B {{7,0,0},{0,0,0},{0,0,1}};
    vector<vector<int>> ans;
    ans = sol.sparseMatrixMultiplication(A,B);
    for(auto r:ans){
        for(auto e:r){
            cout<<e<<" ";
        }
        cout<<endl;
    }
    
    return 0;
}

File: SparseMatrixMultiplication.java
// Time Complexity : O(mnp)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    public int[][] multiply(int[][] mat1, int[][] mat2) {
        if(mat1 == null || mat2 == null) {
            return new int[0][0];
        }
        
        int m = mat1.length;
        int n = mat1[0].length;
        int p = mat2[0].length;
        int[][] result = new int[m][p];
        
        for(int i = 0; i < m; i++) {
            for(int k = 0; k < n; k++) {
                for(int j = 0; j < p; j++) {
                    if(mat1[i][k] != 0) {
                        result[i][j] += mat1[i][k] * mat2[k][j];
                    }
                }
            }
        }
        
        return result;
    }
}